### ISSUE TITLE
[ENH] Benchmarking module with a focus on real-world applications

### ISSUE THREAD

**Issue by tcuongd** [sktime/sktime#4543]
**Is your feature request related to a problem? Please describe.**

Time series packages often provide their own benchmarks (e.g. [Nixtla/statsforecast](https://github.com/Nixtla/statsforecast/blob/main/nbs/examples/AutoArima_vs_Prophet.ipynb), [Uber/orbit](https://www.uber.com/en-AU/blog/orbit/)) using different datasets, time horizons, and metrics. There are very few centralised frameworks and benchmarking studies; probably the most trustworthy is https://forecastingdata.org/ from Monash University, but the datasets aren't grouped in an easily interpretable way and the focus is purely on accuracy. The aim is to have a central benchmarking reference for time series practitioners, similar to how we have https://duckdblabs.github.io/db-benchmark/ to assess DataFrame frameworks.

**Describe the solution you'd like**

1. Build a flexible benchmarking module within sktime. We already have standardised classes for models, hyperparameter optimisation, and metrics, but not for datasets and splitting -- figuring out the datasets interface will be a core part of this step.
2. Include a minimal set of datasets, time horizons, models, and metrics relevant for time series practitioners in industry. Types of time series could include hourly stock tickers, daily internet and consumer goods activity, monthly economic indicators. Time horizons may cover 1 hour to 1 day for hourly datasets, 7 days to 90 days for daily datasets, and 6 to 12 months for monthly datasets. Models could cover traditional ETS models, generalised additive models like Prophet, and neural network architectures, with and without hyperparameter optimisation. Metrics could include validation accuracy, training + prediction runtime, and uncertainty interval coverage.
3. The benchmarking module should also make it easy to filter these defaults, or swap out settings for individual use cases.
4. Finally, we can publish / maintain a page showing the results over the default sets, similar to https://duckdblabs.github.io/db-benchmark/ page.

Ultimately, we want this page to be the first point of reference for time series practitioners, with an option to modify / extend the study to their use case using our benchmarking module.

**Describe alternatives you've considered**

* Focusing on building a very flexible benchmarking module: there's potentially too a high a barrier to entry (or time constraint) for practitioners to learn this API and collect and modify their own custom data, so adoption may not be as a high with this route.
* Focusing on releasing a single benchmarking study: this will get us to something we can show faster, but we inevitably have to add / update datasets and models, so we'll need it to be reproducible anyways. Some time series tasks are also niche in some sense (could be different data, or horizons, or metrics), so allowing practitioners to swap out individual components of the study would make it useful to more people.

**Additional context**

cc @fkiraly @JonathanBechtel

**Comment by fkiraly** [comment]
I think this is an excellent suggestion!

* I think `sktime` as a "neutral ground" with open governance makes sense for this
* imo the ability to re-run and add new estimators for users is key (instead of just re-running the study with a hard coded set of estimators) - it should be easy for anyone to use the tool to run their own "counterstudy" if they feel an estimator has been left out, or the study was badly conducted.
* "very flexible" I agree would probably be overkill, I fully agree we should strike a balance between hard-coded study and fully flexible framework. (e.g., the first version of the `benchmarking` module which imo is over-desigined a bit)

Individuals who may be interested to chip in:

* @ngupta23, who has a benchmarking framework in `pycaret`, although we may like to design it closer to the `scikit-base` intercaces. We could go with a partial refactor that moves some parts of `pycaret` to the lower, `sktime` layer with added flexibility.
* @DBCerigo and @alex-hh, who have a benchmarking framework in `kotsu` and has contributed a `kotsu` based benchmarking framework to `sktime`, see the `benchmarking_v2.ipynb` example. We could move based on that, with dataset and task abstractions.
* @TNTran92, wo has been working on benchmarking during the summer period of 2022

Related PR:
* dataset abstraction based on `scikit-base` https://github.com/sktime/sktime/pull/4333, this could be connected with abstractions for an "experimental study", possibly taking some ideas from `pycaret` or `kotsu`

**Comment by DBCerigo** [comment]
The notebook to checkout for (one of) the working benchmarking core interface, as implemented for forecasting models, is https://github.com/sktime/sktime/blob/main/examples/04_benchmarking_v2.ipynb

Lmk if you have any questions or thoughts on it. 

Some next steps for that way forward would be to implement as per https://github.com/sktime/sktime/blob/main/sktime/benchmarking/forecasting.py is for Forecasting models for Classification, Clustering, etc.

**Comment by ngupta23** [comment]
```
@ngupta23, who has a benchmarking framework in pycaret, although we may like to design it closer to the scikit-base interfaces. We could go with a partial refactor that moves some parts of pycaret to the lower, sktime layer with added flexibility.
```

I built one based on what nixtla had done. You can [find it here](https://github.com/pycaret/benchmarks).

**Comment by fkiraly** [comment]
@ngupta23, this is excellent, I had not seen that!

My first impression is, this has the kind of logic to need, but the extender contract is not flexible.

My gut feeling tells me, we want to push it one level lower, on python, with an extender contract?
Your framework is more of the kind that you would run in a performance monitoring or comparison app in deployment - that's excellent because that's what many people need, and it could use the intermediate, more configurable level?

**Comment by ngupta23** [comment]
> FYI @ngupta23, I have a faint feeling that @tcuongd might like to see Prophet added in the experiments?

I think I did benchmark Prophet in there but maybe not for all versions of pycaret.

> Further, have you built anything for probabilistic metrics?

No, I have not done this.


> My gut feeling tells me, we want to push it one level lower, on python, with an extender contract?
Your framework is more of the kind that you would run in a performance monitoring or comparison app in deployment - that's excellent because that's what many people need, and it could use the intermediate, more configurable level?

I wanted some simple benchmarks to compare the basic functionality across models and also to ensure that future releases of pycaret do not regress in performance. Not a perfect method since we only look at M3 and M4 that too for only a subset of the functions available from pycaret, but it serves our immediate needs. It may not be what you are looking for though for sktime.

**Comment by fkiraly** [comment]
> I wanted some simple benchmarks to compare the basic functionality across models and also to ensure that future releases of pycaret do not regress in performance. Not a perfect method since we only look at M3 and M4 that too for only a subset of the functions available from pycaret, but it serves our immediate needs. It may not be what you are looking for though for sktime.

I see - might I interest you then in designing (or building) sth together more user facing?
Ultimately, you should also be able to use that for your pycaret testing/regression benchmark!

**Comment by DBCerigo** [comment]
> @DBCerigo, @ngupta23, would you be interested in a design round on the specification language, user experience, and extender contract?

I don't have capacity right now for any significant design work. I'm happy to answer any Qs (@ me) on the implementation referred in my comment above.

**Comment by ngupta23** [comment]
> I don't have capacity right now for any significant design work. I'm happy to answer any Qs (@ me) on the implementation referred in my comment above.

Same here. Unfortunately stretched too thin right now with existing commitments, but happy to answer any questions about the implementation in `pycaret`.

**Comment by vortex-wq** [comment]
Hi, @tcuongd I read through your suggested design, if still open can you guide how to move it forward?
